### PR TITLE
AOS2-91 migrated everything to nx project

### PR DESCRIPTION
**Pull request by m-thompson-code**
### Jira

https://bitovi.atlassian.net/browse/AOS2-91

### Description

1. Migrated everything to an Nx workspace
2. Adds tests
3. Updated github actions to publish from tools/eslint-rules

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Editor configuration, see http://editorconfig.org
+root = true
+
+[*]
+charset = utf-8
+indent_style = space
+indent_size = 2
+insert_final_newline = true
+trim_trailing_whitespace = true
```
Comment (fabioemoutinho) [inline .editorconfig:9]: are these nx defaults or personal preference? (just curious)

File: README.md [hunk 2]
```diff
@@ -1,23 +1,48 @@
-# @bitovi/eslint-plugin
+# BitoviEslintPlugin
```
Comment (fabioemoutinho) [inline README.md]: Bitovi ESLint Workspace

**Comment by christopherjbaker** [comment]
LGTM. Couple questions/points:
1. What is the "tools" directory for? If `eslint-rules` is being published, I'd think it would go in `libs`?
2. Since we're setting up nx anyway, would it make sense to bring the eslint-config code into this same repo (maybe rename `eslint-workspace`)? At least for now, I'd keep it publishing to npm the same way, but put all the code in one spot.
3. If 2 is yes, then it probably also make sense (in the future, not now) to publish React and Angular outputs as separate packages (`X-plugin` I guess, so it can have both config and plugins)?
4. There are several `.gitkeep` files that I don't think we need. (I've always hated this pattern, TBH.) The one in apps is already not needed. If eslint-rules goes to libs, that one isn't needed. Then there's one in tools/generators, but if we don't need any generators then do we need the directory?

If you want to do 2 or 3 now but don't have the bandwidth, I can get another set of hands for you.

**Review by fabioemoutinho — APPROVED** [review]
apart from a couple of comments, lgtm

File: .editorconfig (hunk 1 above)
Comment (m-thompson-code) [reply .editorconfig:9]: yupyup, just about everything in this PR is default from Nx

**Comment by m-thompson-code** [comment]
> 

> What is the "tools" directory for? If eslint-rules is being published, I'd think it would go in libs?

First things first, here's information on `eslint-rules`:

`eslint-rules` was generated using the `@nrwl/linter - workspace-rule` plugin.

Under the hood, `eslint-rules` is just an Nx lib. Here are some considerations to why it doesn't live in the traditional `libs` directory.

The tools directory in Nx is typically for generators. When using Nx's `@nrwl/linter - workspace-rule` plugin (which comes with Nx out of the box),
it addes `eslint-rules` to `tools` by default, and adds these rules to `nx.json`:

```json
{
    "targetDefaults": {
        "build": {/* ... */},
        "test": {/* ... */},
        "e2e": {/* ... */},
        "lint": {
            "inputs": [
                "default",
                "{workspaceRoot}/.eslintrc.json",
                "{workspaceRoot}/tools/eslint-rules/**/*"// <-- added when using `@nrwl/linter - workspace-rule`
            ]
        }
    },
}
```

It isn't normal for `eslint-rules` to be "publishable", but since `eslint-rules` is really just some lib, you can add lint and build targets to its `project.json`.

By adding what lint and build targets generated by the `@nrwl/node` plugin, we can lint and build our eslint rules.

> If eslint-rules is being published, I'd think it would go in libs?

For now, I think we'll have to settle for the current directory. `@nrwl/linter - workspace-rule` plugin doesn't give us the freedom to choose. We have future plans to write a wrapper generator so we can more-easily choose where things go, but for now all directories must be nested under `tools/eslint-rules`.

> Since we're setting up nx anyway, would it make sense to bring the eslint-config code into this same repo (maybe rename eslint-workspace)? At least for now, I'd keep it publishing to npm the same way, but put all the code in one spot.

Yea, I think this would make a lot of sense to do. Though for now, I think we should limit the scope of this current effort to just migrating `eslint-plugin` to Nx. We had 2 main goals with this PR:

1. Add typescript to `eslint-plugin`
2. Have a quicker way to generate eslint rules and tests for those rules

> If 2 is yes, then it probably also make sense (in the future, not now) to publish React and Angular outputs as separate packages (X-plugin I guess, so it can have both config and plugins)?

This would make sense. I think there's room for a lot of discussion. My main focus after this PR would be how do we go about writing our generators to support what we want to do moving forward.

> There are several .gitkeep files that I don't think we need. (I've always hated this pattern, TBH.) The one in apps is already not needed. If eslint-rules goes to libs, that one isn't needed. Then there's one in tools/generators, but if we don't need any generators then do we need the directory?

I don't have a hard opinion on these .gitkeeps, they are generated with Nx. It feels like home with them there and without them, I can barely tell this is an Nx workspace. All jokes aside, I think the .gitkeeps are kinda useful since the directories for apps, libs and generators are different depending on what kind of Nx workspace you generate and by keeping them, it's less of a mystery.

Either way, they can be removed though. I just don't care enough about it.

> If you want to do 2 or 3 now but don't have the bandwidth, I can get another set of hands for you.

I'd love to discuss future development on this stuff. Depending on everyone's bandwidth, I think we should create clear goals and write up tickets on what direction we want to go in next.

**Comment by m-thompson-code** [comment]
The only thing left to do is update how we go about publishing. There's a publish npm script, but I assume we want to use the existing pipeline. And I don't know how to go about doing so. @christopherjbaker, what are your thoughts?

File: .editorconfig (hunk 1 above)
Comment (vdtdev) [reply .editorconfig:9]: I think I pulled these from a newly generated Angular app, but I don't remember

**Comment by m-thompson-code** [comment]
> The only thing left to do is update how we go about publishing. There's a publish npm script, but I assume we want to use the existing pipeline. And I don't know how to go about doing so. @christopherjbaker, what are your thoughts?

After discussing with @christopherjbaker, I've updated github actions to publish eslint-rules properly

File: tools/eslint-rules/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@bitovi/eslint-plugin",
-  "version": "2.0.0",
+  "version": "1.0.1",
```
Comment (fabioemoutinho) [inline tools/eslint-rules/package.json:3]: Is this going to be auto bumped?

Comment (m-thompson-code) [reply tools/eslint-rules/package.json:3]: yea, this PR won't be released directly, we'll release through github actions instead

**Review by vdtdev — APPROVED** [review]
lgtm! Love the tests! ;D

**Pull request merged**